### PR TITLE
Issue #SB-7061 feat: User Profile Visibility - Global Settings

### PR DESCRIPTION
**Pull request by JaikumarSRajan**

### CODE REVIEW

File: cassandra-migration/src/main/resources/db/migration/cassandra/V1.24_cassandra.cql [hunk 1]
```diff
@@ -1 +1,3 @@
...
+
+INSERT INTO system_settings(id, field, value) VALUES('userProfileConfig', 'userProfileConfig', '{ "userDataFields": ["firstName", "lastName", "profileSummary", "avatar", "countryCode", "dob", "email", "gender", "grade", "language", "location", "phone", "subject", "userName", "webPages", "jobProfile", "address", "education", "skills", "organisations", "badgeAssertions"], "eternalPublicFields": ["firstName", "lastName", "profileSummary"], "eternalPrivateFields": ["email", "phone"] }');
```
Comment (bvinayakumar) [inline cassandra-migration/src/main/resources/db/migration/cassandra/V1.24_cassandra.cql]: We already ran 1.24 on Sunbird Dev. So, put this insert into V1.25_cassandra.cql.

File: common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java [hunk 2]
```diff
@@ -691,6 +691,13 @@
   public static final String LINK = "link";
   public static final String SET_PW_LINK = "setPasswordLink";
   public static final String SUNBIRD_URL_SHORTNER_ENABLE = "sunbird_url_shortner_enable";
+  public static final String USER_PROFILE_CONFIG = "userProfileConfig";
+  public static final String USER_DATA_FIELDS = "userDataFields";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java]: userDataFields -> fields

File: common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java [hunk 3]
```diff
@@ -691,6 +691,13 @@
...
+  public static final String ETERNAL_PUBLIC_FIELDS = "eternalPublicFields";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java]: eternalPublicFields -> publicFields
eternalPrivateFields -> privateFields
sunbird_default_user_profile_visibility -> sunbird_user_profile_field_default_visibility
defaultUserProfileVisibility -> defaultProfileFieldVisibility

File: common-util/src/main/java/org/sunbird/common/util/ConfigUtil.java [hunk 4]
```diff
@@ -81,4 +81,38 @@ private static Config createConfig(String fileName) {
     Config envConf = ConfigFactory.systemEnvironment();
     return envConf.withFallback(defaultConf);
   }
+
+  /*
+   * @desc This method will parse json string and return config object
+   * @param json string to be parsed
+   * @return config object generated by parsing the json string
+   */
+  public static Config parseJSONString(String jsonString) {
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/util/ConfigUtil.java]: parseJSONString -> getConfigFromJsonString

File: common-util/src/main/resources/externalresource.properties [hunk 5]
```diff
@@ -109,4 +109,5 @@ sunbird_sso_password=
 sunbird_sso_url=
 sunbird_sso_realm=
 sunbird_keycloak_required_action_link_expiration_seconds=155520000
-sunbird_url_shortner_enable=false
\ No newline at end of file
+sunbird_url_shortner_enable=false
+sunbird_default_user_profile_visibility=public
```
Comment (bvinayakumar) [inline common-util/src/main/resources/externalresource.properties]: sunbird_user_profile_field_default_visibility

File: common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java (hunk 2 above)
Comment (JaikumarSRajan) [reply common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java]: @bvinayakumar This key is used to fetch **userDataFields** from **userProfileConfig** in system settings. Should that be renamed to **fields**?

Comment (bvinayakumar) [reply common-util/src/main/java/org/sunbird/common/models/util/JsonKey.java]: yep

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseCode.java [hunk 6]
```diff
@@ -643,6 +645,21 @@
...
+  errorConfigLoadEmptyString(
+      ResponseMessage.Key.ERROR_CONFIG_LOAD_EMPTY_STRING,
+      ResponseMessage.Message.ERROR_CONFIG_LOAD_EMPTY_STRING),
+  errorConfigLoadParseString(
+      ResponseMessage.Key.ERROR_CONFIG_LOAD_PARSE_STRING,
+      ResponseMessage.Message.ERROR_CONFIG_LOAD_PARSE_STRING),
+  errorConfigLoadEmptyConfig(
+      ResponseMessage.Key.ERROR_CONFIG_LOAD_EMPTY_CONFIG,
+      ResponseMessage.Message.ERROR_CONFIG_LOAD_EMPTY_CONFIG),
+  fieldConfiguredInMultipleFieldsLists(
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseCode.java]: errorConflictingFieldConfiguration

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 7]
```diff
@@ -355,6 +356,13 @@
     String NO_EMAIL_RECIPIENTS =
         "Email notification is not sent as the number of recipients is zero.";
     String PARAMETER_MISMATCH = "Mismatch of given parameters: {0}.";
+    String ERROR_CONFIG_LOAD_EMPTY_STRING =
+        "Load configuration failed as empty string is passed as parameter.";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: Loading {0} configuration failed as empty string is passed as parameter.

e.g. {0} = user

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 8]
```diff
@@ -355,6 +356,13 @@
...
+    String ERROR_CONFIG_LOAD_PARSE_STRING = "Load configuration failed due to parsing error.";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: Loading {0} configuration failed due to parsing error.

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 9]
```diff
@@ -355,6 +356,13 @@
...
+    String ERROR_CONFIG_LOAD_EMPTY_CONFIG = "Load configuration failed.";
+    String FIELD_CONFIGURED_IN_MULTIPLE_FIELDS_LISTS =
+        "{0} fields are configured in multiple fields lists.";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: Field {0} in {1} configuration is conflicting in {2} and {3}.

e.g. 
{0} = email
{1} = user
{2} = publicFields 
{3} = privateFields

File: common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 10]
```diff
@@ -355,6 +356,13 @@
...
+    String ERROR_SYSTEM_SETTING_NOT_FOUND = "System Setting not found for id - {0}.";
```
Comment (bvinayakumar) [inline common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: System Setting not found for id: {0}

**Pull request merged**